src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,532 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Linq;
+
+namespace Microsoft.Collections.Extensions
+{
+    /// <summary>
+    /// DictionarySlim<TKey, TValue> is similar to Dictionary<TKey, TValue> but optimized in three ways:
+    /// 1) It allows access to the value by ref replacing the common TryGetValue and Add pattern.
+    /// 2) It does not store the hash code (assumes it is cheap to equate values).
+    /// 3) It does not accept an equality comparer (assumes Object.GetHashCode() and Object.Equals() or overridden implementation are cheap and sufficient).
+    /// </summary>
+    [DebuggerTypeProxy(typeof(Extensions.DictionarySlimDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class DictionarySlim<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey>
+    {
+        // We want to initialize without allocating arrays. We also want to avoid null checks.
+        // Array.Empty would give divide by zero in modulo operation. So we use static one element arrays.
+        // The first add will cause a resize replacing these with real arrays of three elements.
+        // Arrays are wrapped in a class to avoid being duplicated for each <TKey, TValue>
+        private static readonly Entry[] InitialEntries = new Entry[1];
+        // 1-based index into _entries; 0 means empty
+        private int[] _buckets;
+        private Entry[] _entries;
+        // 0-based index into _entries of head of free chain: -1 means empty
+        private int _freeList = -1;
+
+        private struct Entry
+        {
+            public TKey key;
+            public TValue value;
+            // 0-based index of next entry in chain: -1 means end of chain
+            // also encodes whether this entry _itself_ is part of the free list by changing sign and subtracting 3,
+            // so -2 means end of free list, -3 means index 0 but on free list, -4 means index 1 but on free list, etc.
+            public int next;
+        }
+
+        public DictionarySlim()
+        {
+            _buckets = HashHelpers.DictionarySlimSizeOneIntArray;
+            _entries = InitialEntries;
+        }
+
+        public DictionarySlim(int capacity)
+        {
+            capacity = HashHelpers.GetPrime(capacity);
+            _buckets = new int[capacity];
+            _entries = new Entry[capacity];
+        }
+
+        // Drop sign bit to ensure non negative index
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private uint GetHashCode(TKey key) => (uint)key.GetHashCode();
```

Comment:
This should be marked as `static` since the method doesn't use state of a dictionary instance.

Reply:
@AndyAyersMS would that affect perf/inlining in any way? Does JIT know to not bother passing `this` ?

Reply:
This should always inline so it doesn't matter if it's static or not. I'd just make it static for clarity.

I think I'm more worried about whatever impact this may have when `TKey` is a struct, hopefully the JIT will elide the copy but I'm not so sure.

Reply:
Unless the method is inlined the jit must pass all the arguments the ABI demands; it can't peer into the method and discover `this` is not used and then decide not to pass it.

But as @mikedn says trivial methods like this are always inlined when possible and then the jit can realize the inlinee body has no uses of `this` and possibly gain some small advantage from knowing that.

I don't think the jit will be able to elide the struct copy mandated by the call so manually inlining this method may improve perf for struct typed keys.